p2p/libp2p/discovery/hostWithConnectionManagement.go

```diff
@@ -0,0 +1,64 @@
+package discovery
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/ElrondNetwork/elrond-go/p2p/libp2p"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+type hostWithConnectionManagement struct {
+	sharder libp2p.Sharder
+	ConnectableHost
+}
+
+// NewHostWithConnectionManagement returns a host wrapper able to decide if connection initiated to a peer
+// will actually be kept or not
+func NewHostWithConnectionManagement(ch ConnectableHost, sharder libp2p.Sharder) (*hostWithConnectionManagement, error) {
+	if ch == nil {
```

Comment:
why not check.ifNil ? for ch and sharder as well.

Reply:
done for ch. For sharder it is allowed to have a nil implementation until the refactoring is complete (because of the chicken-and-egg problem). Added TODO